okhttp/src/main/java/okhttp3/internal/Util.kt

```diff
@@ -0,0 +1,642 @@
+   * so small, and the likelihood of an intersection so great, that it is not worth the CPU cost of
+   * sorting or the memory cost of hashing.
+   */
+  fun nonEmptyIntersection(
+    comparator: Comparator<String>,
+    first: Array<String>?,
+    second: Array<String>?
+  ): Boolean {
+    if (first == null || second == null || first.isEmpty() || second.isEmpty()) {
+      return false
+    }
+    for (a in first) {
+      for (b in second) {
+        if (comparator.compare(a, b) == 0) {
+          return true
+        }
+      }
+    }
+    return false
+  }
+
+  fun hostHeader(url: HttpUrl, includeDefaultPort: Boolean): String {
+    val host = if (url.host().contains(":"))
+      "[" + url.host() + "]"
+    else
+      url.host()
+    return if (includeDefaultPort || url.port() != HttpUrl.defaultPort(url.scheme()))
+      host + ":" + url.port()
+    else
+      host
+  }
+
+  /**
+   * Returns true if [e] is due to a firmware bug fixed after Android 4.2.2.
+   * https://code.google.com/p/android/issues/detail?id=54072
+   */
+  fun isAndroidGetsocknameError(e: AssertionError): Boolean {
+    return (e.cause != null &&
+        e.message?.contains("getsockname failed") == true)
+  }
+
+  fun indexOf(comparator: Comparator<String>, array: Array<String>, value: String): Int {
+    var i = 0
+    val size = array.size
+    while (i < size) {
+      if (comparator.compare(array[i], value) == 0) return i
+      i++
+    }
+    return -1
+  }
+
+  @Suppress("UNCHECKED_CAST") fun concat(array: Array<String>, value: String): Array<String> {
+    val result = array.copyOf(array.size + 1)
+    result[result.size - 1] = value
+    return result as Array<String>
+  }
+
+  /**
+   * Increments [pos] until [input] is not ASCII whitespace. Stops at [limit].
+   */
+  fun skipLeadingAsciiWhitespace(input: String, pos: Int, limit: Int): Int {
+    for (i in pos until limit) {
+      when (input[i]) {
+        '\t', '\n', '\u000C', '\r', ' ' -> Unit
+        else -> return i
+      }
+    }
+    return limit
+  }
+
+  /**
+   * Decrements [limit] until `input[limit - 1]` is not ASCII whitespace. Stops at [pos].
+   */
+  fun skipTrailingAsciiWhitespace(input: String, pos: Int, limit: Int): Int {
+    for (i in limit - 1 downTo pos) {
+      when (input[i]) {
+        '\t', '\n', '\u000C', '\r', ' ' -> Unit
+        else -> return i + 1
+      }
+    }
+    return pos
+  }
+
+  /** Equivalent to `string.substring(pos, limit).trim()`.  */
+  fun trimSubstring(string: String, pos: Int, limit: Int): String {
+    val start = skipLeadingAsciiWhitespace(string, pos, limit)
+    val end = skipTrailingAsciiWhitespace(string, start, limit)
+    return string.substring(start, end)
+  }
+
+  /**
+   * Returns the index of the first character in [input] that contains a character in [delimiters].
+   * Returns limit if there is no such character.
+   */
+  fun delimiterOffset(input: String, pos: Int, limit: Int, delimiters: String): Int {
```

Comment:
Wants to be like, `String.indexOfElement()` I think

Reply:
left these 3 for follow up